### DISCUSSION TITLE
linear FEniCS StationaryModel

### DISCUSSION THREAD

**Discussion by meigel** [pymor/pymor#1262] [Q&A] [answered]
Is it possible to implement a linear StationaryModel with the provided FEniCS bindings? The examples only show nonlinear problems and FenicsOperator requires the operator form to only have one argument.
Cheers, M

**Comment by HenKlei** [comment]
Hi Martin,

yes, that should be possible. For instance if you change the line
```
    F = df.inner((1 + c*u**2)*df.grad(u), df.grad(v))*df.dx - f*v*df.dx
```
in the `fenics_nonlinear.py` (in pymordemos) to something like
```
    F = df.inner(df.grad(u), df.grad(v))*df.dx - f*v*df.dx
```
you obtain the weak formulation of a simple Poisson problem.
I did not try that out, but it should work like this.

Does that answer your question?

**Reply by meigel** [reply]
Thanks for your answer! That certainly would work. However, I would like to invoke a linear solver and avoid Newton iterations. I infer from your suggestion that there is no direct way to use `FenicsOperator` and some other available FEniCS wrapper to construct a `StationaryModel`.

**Comment by renefritze** [comment, marked as answer]
Hey @meigel,

this is quite possible, yes. 
Have a look at the `thermalblock_simple` [demo](https://docs.pymor.org/2020.2.0/_modules/pymordemos/thermalblock_simple.html#discretize_fenics). It's a bit involved due to the per-block operators and such, but should be essentially what you want.

**Reply by meigel** [reply]
I see, thanks! This approach however wraps the already assembled operator, hence precluding a parameter dependent solve, if I get it correctly.

**Reply by renefritze** [reply]
The thermalblock problem is parameter separable, so the parameter dependency is "moved out" from the block operators into per-block `ParameterFunctionals` ([more on that in the tutorial](https://docs.pymor.org/2020.2.0/tutorial_builtin_discretizer.html#a-parametric-problem)). Maybe you can say what kind of problem you are trying to solve so we can see if it needs to be tackled differently?

**Reply by meigel** [reply]
Sure, thanks. I would need a parametric diffusion coefficient of the (generic) form
`kappa = df.Expression('A0 + A1 + A2', A0=1, A1=0, A2=0, degree=2)`
in a standard linear stationary diffusion operator
`a = kappa * df.inner(df.grad(u), df.grad(v)) * df.dx`
My (uneducated) impression is that the mentioned approach cannot be transferred to this, can it?

**Reply by renefritze** [reply]
Maybe I'm missing something, but that looks like a separable setup and if you map that `df.Expression` to a `pymor.parameters.functionals.ExpressionParameterFunctional` ([docs](https://docs.pymor.org/2020.2.0/generated/pymor.parameters.html#pymor.parameters.functionals.ExpressionParameterFunctional)) you'd be mostly good to go.

**Reply by meigel** [reply]
Ah, you are of course right. I should have posted a common non-separable coefficient...
`kappa = df.Expression('exp(A0 + A1 + A2)', A0=1, A1=0, A2=0, degree=2)`

**Reply by sdrave** [reply]
Hi @meigel, that would still work: your coefficient kappa is still a (parameter-dependent) constant that can be pulled out of the integral. As long as the Ai do not depend on x, you can use an `ExpressionParameterFunctional`.

**Reply by meigel** [reply]
Thanks again and sincere apologies for not providing an appropriate example in the first place. In fact, the expression (as used in random PDEs) usually does depend on space, i.e.,
`kappa = df.Expression('exp(A0 + sin(A1*x[0]) + sin(A2*x[1]))', A0=1, A1=0, A2=0, degree=2)`
This now really is the final iteration concerning this function... ;)

**Reply by renefritze** [reply]
In this case, yes, you'd need a different approach to the reduction process. The support for empirical operator interpolation implemented in pyMOR should actually also work with your problem. You've seen the more complicated treatment of nonlinear+non-separable, but linear+non-separable is also a use case. I do not think we have showcase for that though.
@sdrave will have more to say on this.

**Reply by sdrave** [reply]
So I guess what you really want here is empirical interpolation of the data function `kappa`. Unfortunately, there is no official support for that yet in pyMOR, however, you should be able to basically follow the approach from [here](https://github.com/pymor/pymor/discussions/1230#discussioncomment-311657). As you are working with fenics, you would need to adapt the code to obtain `training_data` and the resulting `ParameterFunctionals` with appropriate fenics code. I would need to check, but I assume that this should be relatively straightforward. I am somewhat short on time at the moment, but if you need more help, please do not hesitate to ask. To somewhat push this feature on the agenda, I have also opened a new issue in #1265.

**Reply by meigel** [reply]
Thanks for the suggestion(s)! I think this will work. Adjusting the mentioned code shouldn't be difficult.
